### ISSUE TITLE
End time of 15min counts incorrect: end time sets when survey submited for one user

### ISSUE THREAD

**Issue by CrisSevilleja** [BiologicalRecordsCentre/ABLE#687]
### Describe the issue

A user (user ID 376708) of the app had reported a problem with the incorrect end time of his records (15min counts) on the website (see screenshot 1). At first, I thought it was related to the UK time of the server, but the difference is so big that it cannot be. 

It looks like the end time is not created in relation to the end of the survey, but rather the time of data submission. 

This has always happened with this recorder as you can see in the submission. I confirm that the user clicked the "Finish" button at the end of the survey, which is the button that sets the end time, and later submitted the surveys at home after checking the images.

15min count samples with this error: ID25426480, ID25235815 (screenshots)

I don't know if the error comes from the app or the website, the times cannot be checked on the app after submission. @kazlauskis @andrewvanbreda 

### Expected behaviour

Store the End Time of the 15min counts and visualize it correctly on the website https://butterfly-monitoring.net/mydata/samples

### To Reproduce

1. Do a 15min count
2. Click on Finish
3. Click on Upload after checking the survey
4. go to My samples page on eBMS website https://butterfly-monitoring.net/mydata/samples and see Start and End time of the 15min count correct as recorded

### App Version

v1.26.3 (350)

### Device

iPhone 11 version IOS 17.3.1.

### When did this happen

since this user is registering, this year

### Screenshots

![1 error on 15min count times user 376708](https://github.com/BiologicalRecordsCentre/ABLE/assets/25052478/e16cc0ff-f941-4a86-8f94-27a0e9b44534)
![ID25426480](https://github.com/BiologicalRecordsCentre/ABLE/assets/25052478/e6d4eff3-0536-46af-9268-b05f7977e8b0)
![ID25235815](https://github.com/BiologicalRecordsCentre/ABLE/assets/25052478/dc4935de-6caa-499e-bce1-f42b9fed86f2)

**Comment by kazlauskis** [comment]
@CrisSevilleja, thanks for filling in the details. This is not a website-related issue, as it is a custom attribute that the website doesn't modify. We set the survey time in the app only in one place, and that is when users tap the "Finish" button. Are you sure the user finishes the survey after 15 minutes and not when they upload it? Unfinished surveys have a 'draft' status, so that would be easy to check.

A test they could run:
1. create a new survey
2. _(take a screenshot, so I can see the time)_
3. wait for 15 minutes
4. _(take a screenshot)_
5. press Finish
6. _(take a screenshot)_
7. wait for another 15 minutes
8. press Upload
9. _(take a screenshot)_
10. send me the sceenshots

**Comment by CrisSevilleja** [comment]
Hi Karolis, thanks for your reply. I asked him directly about pressing the Finish button, but I only had an email conversation about it... I will ask him to do the test and send me all the screenshots. Anyway, it is not possible to submit the surveys without clicking the Finish button, is it?

**Comment by kazlauskis** [comment]
You're right. Users need to press the Finish button before pressing Upload. But they can start a survey, wait for 15 minutes to pass, go home, and then press Finish+Upload later in the evening, so the end time would be that of when it was pressed.

**Comment by CrisSevilleja** [comment]
indeed, that's the point to discover, if they press finish just before submitting at home or correctly on the field. I already asked, let's see what we get. thanks!

**Comment by CrisSevilleja** [comment]
Hi @kazlauskis the user replied with the screenshots. Apparently, he is doing correctly the submission of the 15min count, by clicking the Finish button. I mentioned again the importance of clicking the Finish button at the end of the count and not when submitting and he said he did it correctly. 

The test you asked for:
1. create a new survey
![1](https://github.com/BiologicalRecordsCentre/ABLE/assets/25052478/c9b7ca56-6f50-429f-9bd7-60bc4045065a)
2. wait for 15 minutes 
![02](https://github.com/BiologicalRecordsCentre/ABLE/assets/25052478/fc322f31-931e-46f2-990a-457cdda0aee2)
3. press Finish
![03](https://github.com/BiologicalRecordsCentre/ABLE/assets/25052478/a97fac51-5c90-4f6c-8c1a-b4cb687e2f5a)
4. wait for another 15 minutes & press Upload (he submitted the survey)
![04](https://github.com/BiologicalRecordsCentre/ABLE/assets/25052478/86341943-4226-43f9-b50a-18a3616b0443)

Checking his account (user ID 376708), the surveys submitted recently are not on the website, neither the test nor the ones from the 30 of May... There is something on the connection between the website and the app as Chris mentioned #686 
![5](https://github.com/BiologicalRecordsCentre/ABLE/assets/25052478/35efad44-8366-49e4-9686-be7f6fad7aca)

I hope this was useful.

**Comment by kazlauskis** [comment]
@CrisSevilleja thanks, the test shows the app works OK in this example. I can see the record in the app and on the website with the finish time of 17:58. I'll keep an eye on this, but at the moment, I don't see why records could have such finish times, so most likely, users press Finish when uploading.

<img width="849" alt="Screenshot 2024-06-10 at 12 09 43" src="https://github.com/BiologicalRecordsCentre/ABLE/assets/7383162/53bc08ab-5717-47b4-a6b5-72f6e20e2c2f">

Just stepping back for a moment, please remind me why we must capture the end time. At the moment, it just sets a timestamp, that's all. The GPS isn't affected as it is stopped when the 15 minutes run out, and the app still allows adding species beyond pressing the Finish, so it is not that we know when the user has actually stopped surveying.

**Comment by CrisSevilleja** [comment]
Thank you @kazlauskis, I will communicate this to the user, It's a bit strange but the most probable case is he is not pressing well the button after finishing the survey. 

We are capturing the End Time because it is not always 15min in case the recorder pauses the survey. Then we can filter the surveys with longer or shorter times than 15min. I think that is the case, maybe @chrisvanswaay or @DavidRoy can corroborate it. However, as you mentioned, it depends on the correct behaviour of the recorder pressing the Finish button at the end of the survey and we cannot be sure they pressed it at the correct time. Maybe we can place a short note under the end time mentioning to press the End time after finishing the survey...

**Comment by CrisSevilleja** [comment]
Hi @kazlauskis Discussing this with @DavidRoy  we agreed that the best way to avoid these types of issues is to determine the end time when the GPS ends, i.e. when the last GPS reading is taken and the 15-minute countdown ends. This way, the end time is not affected by whether the Finish button is clicked or not. 

It would be great to check which 15minute counts have excessive recording times (more than 20–30 minutes) and correct them, or contact the user depending on the number.

**Comment by kazlauskis** [comment]
@CrisSevilleja I can set the end time when the 15-minute timer runs out, but this doesn't guarantee that the survey was completed in 15 minutes. If a user starts a survey, but then moves away from the app for 1 hour, then the app will set the survey end time when they return back to the app, and the total survey time will be 1 hour. 

If the total time exceeds 15 minutes we can show a warning and the end time field. Users will be able to manually correct the time.

**Comment by CrisSevilleja** [comment]
Yes, I understand there will be surveys with more than 15minutes if they use the pause several times. The problem we got was that some people didn't press the "finnish" button to complete the survey and determine the End time much longer, even several hours when that was not the real survey at all. 

I wouldn't like to set a warning after 15minutes because we allow to use the pause button and I think it is good to allow people see individuals better for identification, take a photo or capture the butterfly. But maybe we can set up a warning after 20-30 minutes to remind finnishing the survey. 

Does that make more sense? @DavidRoy do you agree?

**Comment by DavidRoy** [comment]
Thanks @kazlauskis @CrisSevilleja.  After discussing the issue and re-reading this thread, I agree with @CrisSevilleja suggestion to set the end time when the 15 minute timer finishes. If the function of the 'Finish' button is therefore to comfirm completion, I wonder if we should change this to 'Finish and upload' to remove one additional upload step?

**Comment by chrisvanswaay** [comment]
When I analyse data I calculate the length of the track and the time between start and end, and if this is >900 seconds I put it at 900 seconds. With this I calculate speed and where speed is >5 m/s I remove. So far I only work with numbers per 1000m track. This procedure removes most of the outliers.

**Comment by DavidRoy** [comment]
That's great to know you have analysis steps to deal with this issue - I like this approach. 

@kazlauskis  implementing the end-time = timer-end will also help avoid confusion from users (i.e. why this issue was raised originally) who expect the end time to be when the 15 minutes (surveying time) runs out.

**Comment by kazlauskis** [comment]
OK, I will automatically set the end time when the 15 minutes are up. The user will still be able to pause the timer and also move away from the app, so the total time might be more than 15 minutes. The main difference from the current implementation will be that users won't need to press the "Finish" button to set the end time, it will be done automatically.

As per [#808](https://github.com/BiologicalRecordsCentre/ABLE/issues/808), once the survey end time is set, I will show it in the app. I will also show the start time to keep it consistent. 

Should we allow changing the end time?

**Comment by CrisSevilleja** [comment]
Everything sounds good, the end time will be set autommatically now and yes, there could be surveys with longer times than 15 minutes. 

I think it would be good to allow changing the End time on the app in case there are issues with the GPS.

David mentioned before to remove the Finnish button and merge with the Upload but I think everyone is used to and this modification won't be appreciated by users. I would keep as we are now with the two steps, first click on Finish and then Upload.

**Comment by kazlauskis** [comment]
I have completed this. The app checks if the survey timer has ended and then sets the end time accordingly. 

A few exceptions. This won't work if the user restarts the app or device. In such cases, pressing the Finish button will set the end time as it used to do. 

Also, if they keep their device locked when the 15-minute timer runs out, it might not set the end time right away, but it will when the user opens the app again.
